pkg/pipeline/ingest/ingest_kafka.go

```diff
@@ -0,0 +1,172 @@
+const defaultBatchReadTimeout = int64(100)
+
+// Ingest ingests entries from kafka topic
+func (r *ingestKafka) Ingest(process ProcessFunction) {
+	// initialize background listener
+	r.kafkaListener()
+
+	// forever process log lines received by collector
+	r.processLogLines(process)
+
+}
+
+// background thread to read kafka messages; place received items into ingestKafka input channel
+func (r *ingestKafka) kafkaListener() {
+	log.Debugf("entering  kafkaListener")
+
+	go func() {
+		var kafkaMessage kafkago.Message
+		var err error
+		for {
+			// block until a message arrives
+			log.Debugf("before ReadMessage")
+			kafkaMessage, err = r.kafkaReader.ReadMessage(context.Background())
+			if err != nil {
+				log.Errorln(err)
+			}
+			log.Debugf("string(kafkaMessage) = %s\n", string(kafkaMessage.Value))
+			if len(kafkaMessage.Value) > 0 {
+				r.in <- string(kafkaMessage.Value)
+			}
+		}
+	}()
+
+}
+
+// read items from ingestKafka input channel, pool them, and send down the pipeline
+func (r *ingestKafka) processLogLines(process ProcessFunction) {
+	var records []interface{}
+	duration := time.Duration(r.kafkaParams.BatchReadTimeout) * time.Millisecond
+	for {
+		select {
+		case <-r.exitChan:
+			log.Debugf("exiting ingestKafka because of signal")
+			return
+		case record := <-r.in:
+			records = append(records, record)
+		case <-time.After(duration): // Maximum batch time for each batch
+			// Process batch of records (if not empty)
+			if len(records) > 0 {
+				process(records)
+				r.prevRecords = records
+				log.Debugf("prevRecords = %v", r.prevRecords)
+			}
+			records = []interface{}{}
+		}
+	}
+}
+
+// NewIngestKafka create a new ingester
+func NewIngestKafka() (Ingester, error) {
+	log.Debugf("entering NewIngestKafka")
+	ingestKafkaString := config.Opt.PipeLine.Ingest.Kafka
+	log.Debugf("ingestKafkaString = %s", ingestKafkaString)
+	var jsonIngestKafka api.IngestKafka
+	err := json.Unmarshal([]byte(ingestKafkaString), &jsonIngestKafka)
+	if err != nil {
+		return nil, err
+	}
+
+	// connect to the kafka server
+	startOffsetString := jsonIngestKafka.StartOffset
+	var startOffset int64
+	switch startOffsetString {
+	case "FirstOffset", "":
+		startOffset = kafkago.FirstOffset
+	case "LastOffset":
+		startOffset = kafkago.LastOffset
+	default:
+		startOffset = kafkago.FirstOffset
+		log.Errorf("illegal value for StartOffset: %s\n", startOffsetString)
+	}
+	log.Debugf("startOffset = %v", startOffset)
+	groupBalancers := make([]kafkago.GroupBalancer, 0)
+	for _, gb := range jsonIngestKafka.GroupBalancers {
+		switch gb {
+		case "range":
+			groupBalancers = append(groupBalancers, &kafkago.RangeGroupBalancer{})
+		case "roundRobin":
+			groupBalancers = append(groupBalancers, &kafkago.RoundRobinGroupBalancer{})
+		case "rackAffinity":
+			groupBalancers = append(groupBalancers, &kafkago.RackAffinityGroupBalancer{})
+		default:
+			return nil, fmt.Errorf("the provided kafka balancer is not valid: %s", gb)
```

Comment:
@KalmanMeth Maybe we need to set default if there was no input from the user?

Reply:
OK